pyanaconda/ui/cockpit/__init__.py

```diff
@@ -0,0 +1,112 @@
+# Public License for more details.  You should have received a copy of the
+# GNU General Public License along with this program; if not, write to the
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+
+import time
+
+from pyanaconda import ui
+from pyanaconda.core.constants import QUIT_MESSAGE
+from pyanaconda.core.util import startProgram
+
+from pyanaconda.anaconda_loggers import get_module_logger
+log = get_module_logger(__name__)
+
+
+class CockpitUserInterface(ui.UserInterface):
+    """This is the main class for Cockpit user interface."""
+
+    def __init__(self, storage, payload,
+                 productTitle="Anaconda", isFinal=True,
+                 quitMessage=QUIT_MESSAGE):
+        """
+        For detailed description of the arguments see
+        the parent class.
+
+        :param storage: storage backend reference
+        :type storage: instance of pyanaconda.Storage
+
+        :param payload: payload (usually dnf) reference
+        :type payload: instance of payload handler
+
+        :param productTitle: the name of the product
+        :type productTitle: str
+
+        :param isFinal: Boolean that marks the release
+                        as final (True) or development
+                        (False) version.
+        :type isFinal: bool
+
+        :param quitMessage: The text to be used in quit
+                            dialog question. It should not
+                            be translated to allow for change
+                            of language.
+        :type quitMessage: str
+        """
+
+        super().__init__(storage, payload)
+
+        self.productTitle = productTitle
+        self.isFinal = isFinal
+        self.quitMessage = quitMessage
+
+    def setup(self, data):
+        """Construct all the objects required to implement this interface.
+
+        This method must be provided by all subclasses.
+
+        TODO: start cockpit backend stuff, unless thats started elsewhere
+        """
+        pass
+
+    def run(self):
+        """Run the interface."""
+        log.debug("web-ui: starting cockpit web view")
+        startProgram(["/usr/libexec/cockpit-desktop",
+                     "/cockpit/@localhost/anaconda-webui/index.html"],
+                     reset_lang=False)
+        log.debug("cockpit web view has been started")
+        # at the moment all the other components expect the UI run() method to be
+        # blocking, so make sure this is the case for the Web UI as well
+        # FIXME: make it possible to stop blocking once Web UI supports shutdown
+        log.debug("web-ui: cockpit web view has been started")
+        log.debug("web-ui: blocking the run() method")
+        time.sleep(100000000)
```

Comment:
I think I have already suggested to use:

```
proc = startProgram(...)
proc.wait()
```

It didn't work?

Reply:
I either missed it or forgot. :P Seems to work fine & is much cleaner, thanks! :)

Reply:
Hmm, it is still sleeping.

Reply:
Ooops, should be foxed for real now. :)